core/DataAccess/LogAggregator.php

```diff
@@ -359,26 +389,40 @@ public function generateQuery($select, $from, $where, $groupBy, $orderBy, $limit
             $logTablesProvider = $this->getLogTableProvider();
             $logTablesProvider->setTempTable(new LogTableTemporary($segmentTable));
 
-            foreach ($logTablesProvider->getAllLogTables() as $logTable) {
-                if ($logTable->getDateTimeColumn()) {
-                    $whereTest = $this->getWhereStatement($logTable->getName(), $logTable->getDateTimeColumn());
-                    if (strpos($where, $whereTest) === 0) {
-                        // we don't need to apply the where statement again as it would have been applied already
-                        // in the temporary table... instead it should join the tables through the idvisit index
-                        $where = ltrim(str_replace($whereTest, '', $where));
-                        if (stripos($where, 'and ') === 0) {
-                            $where = substr($where, strlen('and '));
+            if (!$prebuiltQuery) {
+
+                foreach ($logTablesProvider->getAllLogTables() as $logTable) {
+                    if ($logTable->getDateTimeColumn()) {
+                        $whereTest = $this->getWhereStatement($logTable->getName(), $logTable->getDateTimeColumn());
+                        if (strpos($where, $whereTest) === 0) {
+                            // we don't need to apply the where statement again as it would have been applied already
+                            // in the temporary table... instead it should join the tables through the idvisit index
+                            $where = ltrim(str_replace($whereTest, '', $where));
+                            if (stripos($where, 'and ') === 0) {
+                                $where = substr($where, strlen('and '));
+                            }
+                            $bind = array();
+                            break;
                         }
-                        $bind = array();
-                        break;
                     }
                 }
 
+            } else {
+
+                // Prebuilt queries must include the idvisit field in the select clause for this wrapper to work
+                $prebuiltQuery = 'SELECT sub.* FROM (' . $prebuiltQuery . ') AS sub ' .
```

Comment:
fyi will this one work when config flag is `enable_segments_cache = 0`?

Reply:
Good spot :+1: With `enable_segments_cache = 0` no segment would be applied for prebuilt queries as `getSelectQuery()` isn't used. Do you think it would be acceptable to always use a temporary table for prebuilt query segments even with `enable_segments_cache = 0` ? 

#14776 suggests we might remove this option in Matomo 5.0?

Reply:
@tsteur The code used for page goals archiving now simply does not longer look at the config `enable_segments_cache`, so a temporary table is always used in that case. Imho that should be fine to be already included in Matomo 4. Even though we might discontinue support for the config flag only as of Matomo 5. Any further thoughts on that?

Reply:
Generally I think we won't discontinue the support for that feature in Matomo 5 but we could think about not supporting it as part of this feature. 

If the feature is disabled, I wonder if it would be better to 
* Not archive data (wouldn't show any data) which would look buggy in the UI and we'd need to explain why they are seeing no data
* Try to archive data (would show "wrong" data or if they have it disabled because temporary tables aren't supported by their DB or User permissions then it could maybe throw an error). Can look buggy too and we'd need to explain it also.

Here's another thought how we solved this in MultiChannelConversionAttribution without needing to change `generateQuery` see https://github.com/innocraft/plugin-MultiChannelConversionAttribution/blob/4.1.1/Archiver.php#L134-L157 . It should work here too (but haven't tested):

There's the inner query:

```sql
SELECT
              log_conversion.idvisit,
              COUNT(*) AS num_total
            FROM %slog_conversion AS log_conversion FORCE INDEX(index_idsite_datetime)
            RIGHT JOIN %slog_link_visit_action AS log_vpast ON log_conversion.idvisit = log_vpast.idvisit
            LEFT JOIN %slog_action lac_past ON log_vpast.".$linkField." = lac_past.idaction
            WHERE log_conversion.server_time >= ?
              AND log_conversion.server_time <= ?
              AND log_conversion.idsite IN (?)
              AND log_conversion.idgoal = ".$idGoal."
              AND log_vpast.server_time <= log_conversion.server_time
              AND lac_past.type = ".($linkField == 'idaction_url' ? '1' : '4')."
            GROUP BY log_conversion.idvisit
            ORDER BY NULL
```

We could generate this query maybe via:
 
```php
$query = ...->generateQuery('log_conversion.idvisit, count(*) as num_total', $from = ['log_conversion', ['table' => 'log_link_visit_action', 'join' => 'RIGHT JOIN'], ...)
``` 

Then we have the segmented query and we could put it in the right place like:

```php
$sql = '... 1 / num_total * lvcon.revenue_shipping AS revenue_shipping,
            1 / num_total * lvcon.revenue_discount AS revenue_discount,
            1 / num_total * lvcon.items AS items
          FROM ( . $query['sql']) . ' ) AS r
          LEFT JOIN %slog_conversion lvcon O...'
```

It would also be partially buggy when using `enable_segment_cache=0` as some actions would be filtered out but it would be equally buggy and causing issues with current solution and this way we could avoid all the special cases maybe and keep code base and tests simple. 

Just a thought.

Reply:
@bx80 @sgiehl have you had a chance to give this a try? Noticed the comment became outdated but wanted to make sure we're giving it a try before adding the additional complexity.

Reply:
@tsteur Building the inner query with `generateQuery()` to get the segment and then concatenating it into the full complex query seems to work :+1: 

I've updated this PR with this new approach and removed the `generateQueryCustom()` method. I've kept the changes to `generateQuery()` which move reusable functionality to private sub-methods (`createSegmentTable()` and `addOriginHintToQuery()`)